File: js/raw/detect_folder.js
/**
 * Resemble AI -- Batch Deepfake Detection with Secure Upload
 *
 * Walk a folder of local media files, upload each via the Secure Upload API,
 * run deepfake detection against the returned media token, and poll until each
 * job reaches a terminal state. Uploads and polling run concurrently so
 * batches of dozens/hundreds of files finish in practical wall-clock time
 * (and well within the 1-hour secure-upload token expiration).
 *
 * Non-media files and subdirectories are skipped. Per-file failures are
 * collected and reported in the final summary; the script never aborts early.
 *
 * Prerequisites:
 *     Node.js 18+ (uses native fetch, FormData, Blob)
 *
 * Usage:
 *     export RESEMBLE_API_KEY="your_api_key"
 *     node detect_folder.js <absolute-folder-path> [output-json-path]
 *
 * If [output-json-path] is omitted, results are written to `<folder>/results.json`.
 *
 * Find your API key at https://app.resemble.ai/hub/api
 */

const fs = require("fs");
const path = require("path");

// ── Configuration ────────────────────────────────────────────────────────────

const API_KEY = process.env.RESEMBLE_API_KEY || "";
const BASE_URL = "https://app.resemble.ai/api/v2";
const MAX_WORKERS = 4; // tune to your rate limit / upload bandwidth

// Extend with any audio/video/image extension you care about.
const MEDIA_EXTS = new Set([".wav", ".mp3", ".mp4", ".mov", ".png", ".jpg", ".jpeg"]);

// ── Helpers ──────────────────────────────────────────────────────────────────

function jsonHeaders() {
  return { Authorization: `Bearer ${API_KEY}`, "Content-Type": "application/json" };
}

function authOnlyHeaders() {
  // For multipart/form-data, fetch sets Content-Type (with boundary) itself.
  return { Authorization: `Bearer ${API_KEY}` };
}

async function pollForResult(uuid, { timeout = 600000, interval = 5000 } = {}) {
  const url = `${BASE_URL}/detect/${uuid}`;
  const deadline = Date.now() + timeout;
  while (Date.now() < deadline) {
    const resp = await fetch(url, { headers: jsonHeaders() });
    if (!resp.ok) {
      throw new Error(`GET /detect/${uuid} returned ${resp.status}: ${await resp.text()}`);
    }
    const { item } = await resp.json();
    const status = (item && item.status) || "unknown";
    if (status === "completed" || status === "failed") return item;
    await new Promise((r) => setTimeout(r, interval));
  }
  throw new Error(`polling timed out after ${timeout / 1000}s`);
}

function iterMediaFiles(folder) {
  return fs
    .readdirSync(folder)
    .sort()
    .map((name) => path.join(folder, name))
    .filter((full) => {
      try {
        return fs.statSync(full).isFile() && MEDIA_EXTS.has(path.extname(full).toLowerCase());
      } catch {
        return false;
      }
    });
}

// ── Step 1: Secure upload ────────────────────────────────────────────────────

async function secureUpload(filePath) {
  const buffer = fs.readFileSync(filePath);
  const form = new FormData();
  form.append("file", new Blob([buffer]), path.basename(filePath));

  const resp = await fetch(`${BASE_URL}/secure_uploads`, {
    method: "POST",
    headers: authOnlyHeaders(),
    body: form,
  });
  if (!resp.ok) {
    throw new Error(`POST /secure_uploads returned ${resp.status}: ${await resp.text()}`);
  }
  const data = await resp.json();
  if (!data.token) throw new Error(`no token in secure upload response: ${JSON.stringify(data)}`);
  return data.token;
}

// ── Step 2: Submit detect using the media token ─────────────────────────────

async function submitDetect(mediaToken) {
  const payload = {
    media_token: mediaToken,
    // Prefer webhooks over polling for large batches:
    // callback_url: "https://your-server.example.com/resemble-webhook",
  };
  const resp = await fetch(`${BASE_URL}/detect`, {
    method: "POST",
    headers: jsonHeaders(),
    body: JSON.stringify(payload),
  });
  if (!resp.ok) {
    throw new Error(`POST /detect returned ${resp.status}: ${await resp.text()}`);
  }
  const data = await resp.json();
  const uuid = data && data.item && data.item.uuid;
  if (!uuid) throw new Error("no uuid in detect response");
  return uuid;
}

// ── Step 3: Process one file ─────────────────────────────────────────────────

async function processFile(filePath) {
  const token = await secureUpload(filePath);
  const uuid = await submitDetect(token);
  const result = await pollForResult(uuid);
  const status = (result && result.status) || "unknown";
  if (status !== "completed") {
    throw new Error(`detect job ${uuid} ended with status=${status}`);
  }
  return { file: filePath, detect: result };
}

// ── Concurrency ──────────────────────────────────────────────────────────────

async function runWithConcurrency(items, limit, fn) {
  const succeeded = [];
  const failed = [];
  const iter = items.entries();
  const worker = async () => {
    for (const [, item] of iter) {
      const name = path.basename(item);
      try {
        const record = await fn(item);
        succeeded.push(record);
        const metrics = (record.detect && record.detect.metrics) || {};
        console.log(
          `  [OK]  ${name}  label=${metrics.label ?? null}  score=${metrics.aggregated_score ?? null}`,
        );
      } catch (err) {
        failed.push({ file: item, error: err.message });
        console.log(`  [ERR] ${name}  ${err.message}`);
      }
    }
  };
  await Promise.all(Array.from({ length: limit }, worker));
  return { succeeded, failed };
}

// ── Main ─────────────────────────────────────────────────────────────────────

async function main() {
  if (!API_KEY) {
    console.error("Error: set RESEMBLE_API_KEY environment variable before running.");
    process.exit(1);
  }
  if (process.argv.length !== 3 && process.argv.length !== 4) {
    console.error(
      `Usage: node ${path.basename(process.argv[1])} ` +
        `<absolute-folder-path> [output-json-path]`,
    );
    process.exit(1);
  }
  const folder = process.argv[2];
  if (!path.isAbsolute(folder)) {
    console.error(`Error: folder path must be absolute (got: ${folder}).`);
    process.exit(1);
  }
  let stat;
  try {
    stat = fs.statSync(folder);
  } catch {
    console.error(`Error: not a directory: ${folder}`);
    process.exit(1);
  }
  if (!stat.isDirectory()) {
    console.error(`Error: not a directory: ${folder}`);
    process.exit(1);
  }

  const outputPath = process.argv[3] || path.join(folder, "results.json");

  const files = iterMediaFiles(folder);

  console.log("Resemble AI -- Batch Deepfake Detection with Secure Upload\n");
  console.log("=".repeat(60));
  console.log(`Folder:  ${folder}`);
  console.log(`Output:  ${outputPath}`);
  console.log(`Files:   ${files.length}`);
  console.log(`Workers: ${MAX_WORKERS}\n`);

  if (files.length === 0) {
    console.log("No media files to process.");
    return;
  }

  const { succeeded, failed } = await runWithConcurrency(files, MAX_WORKERS, processFile);

  fs.writeFileSync(outputPath, JSON.stringify({ folder, succeeded, failed }, null, 2));

  console.log();
  console.log("=".repeat(60));
  console.log(
    `Done. ${succeeded.length} succeeded, ${failed.length} failed (of ${files.length}).`,
  );
  console.log(`Results written to ${outputPath}`);
}

main().catch((err) => {
  console.error(`Error: ${err.message}`);
  process.exit(1);
});

File: python/raw/detect_folder.py
"""
Resemble AI -- Batch Deepfake Detection with Secure Upload

Walk a folder of local media files, upload each via the Secure Upload API,
run deepfake detection against the returned media token, and poll until each
job reaches a terminal state. Uploads and polling run concurrently so batches
of dozens/hundreds of files finish in practical wall-clock time (and well
within the 1-hour secure-upload token expiration).

Non-media files and subdirectories are skipped. Per-file failures are
collected and reported in the final summary; the script never aborts early.

Prerequisites:
    pip install requests

Usage:
    export RESEMBLE_API_KEY="your_api_key"
    python detect_folder.py <absolute-folder-path> [output-json-path]

If `output-json-path` is omitted, results are written to `<folder>/results.json`.

Find your API key at https://app.resemble.ai/hub/api
"""

import json
import os
import sys
import time
from concurrent.futures import ThreadPoolExecutor, as_completed

import requests

# ── Configuration ─────────────────────────────────────────────────────────────

API_KEY = os.environ.get("RESEMBLE_API_KEY", "")
BASE_URL = "https://app.resemble.ai/api/v2"
MAX_WORKERS = 4  # tune to your rate limit / upload bandwidth

# Extend with any audio/video/image extension you care about.
MEDIA_EXTS = {".wav", ".mp3", ".mp4", ".mov", ".png", ".jpg", ".jpeg"}

# ── Helpers ───────────────────────────────────────────────────────────────────


def json_headers() -> dict:
    return {"Authorization": f"Bearer {API_KEY}", "Content-Type": "application/json"}


def auth_only_headers() -> dict:
    # For multipart/form-data, requests sets Content-Type (with boundary) itself.
    return {"Authorization": f"Bearer {API_KEY}"}


def poll_for_result(uuid: str, timeout: int = 600, interval: int = 5) -> dict:
    """Poll GET /detect/{uuid} until the job reaches a terminal state."""
    url = f"{BASE_URL}/detect/{uuid}"
    deadline = time.time() + timeout
    while time.time() < deadline:
        resp = requests.get(url, headers=json_headers())
        if not resp.ok:
            raise RuntimeError(f"GET /detect/{uuid} returned {resp.status_code}: {resp.text}")
        item = resp.json().get("item", {})
        status = item.get("status", "unknown")
        if status in ("completed", "failed"):
            return item
        time.sleep(interval)
    raise RuntimeError(f"polling timed out after {timeout}s")


def iter_media_files(folder: str):
    """Yield absolute paths of media files directly inside `folder`, sorted."""
    for name in sorted(os.listdir(folder)):
        full = os.path.join(folder, name)
        if not os.path.isfile(full):
            continue
        if os.path.splitext(name)[1].lower() not in MEDIA_EXTS:
            continue
        yield full


# ── Step 1: Secure upload ─────────────────────────────────────────────────────


def secure_upload(path: str) -> str:
    """Upload a local file to the Secure Upload API and return its media_token."""
    with open(path, "rb") as f:
        resp = requests.post(
            f"{BASE_URL}/secure_uploads",
            headers=auth_only_headers(),
            files={"file": (os.path.basename(path), f)},
        )
    if not resp.ok:
        raise RuntimeError(f"POST /secure_uploads returned {resp.status_code}: {resp.text}")
    token = resp.json().get("token")
    if not token:
        raise RuntimeError(f"no token in secure upload response: {resp.text}")
    return token


# ── Step 2: Submit detect using the media token ──────────────────────────────


def submit_detect(media_token: str) -> str:
    """Submit a detect job referencing a secure-upload token and return the uuid."""
    payload = {
        "media_token": media_token,
        # Prefer webhooks over polling for large batches:
        # "callback_url": "https://your-server.example.com/resemble-webhook",
    }
    resp = requests.post(f"{BASE_URL}/detect", headers=json_headers(), json=payload)
    if not resp.ok:
        raise RuntimeError(f"POST /detect returned {resp.status_code}: {resp.text}")
    uuid = resp.json().get("item", {}).get("uuid")
    if not uuid:
        raise RuntimeError("no uuid in detect response")
    return uuid


# ── Step 3: Process one file ─────────────────────────────────────────────────


def process_file(path: str) -> dict:
    """Upload, submit, and poll one file. Returns a record with the full detect response."""
    token = secure_upload(path)
    uuid = submit_detect(token)
    result = poll_for_result(uuid)
    status = result.get("status", "unknown")
    if status != "completed":
        raise RuntimeError(f"detect job {uuid} ended with status={status}")
    return {"file": path, "detect": result}


# ── Main ──────────────────────────────────────────────────────────────────────


def main():
    if not API_KEY:
        sys.exit("Error: set RESEMBLE_API_KEY environment variable before running.")
    if len(sys.argv) not in (2, 3):
        sys.exit(
            f"Usage: python {os.path.basename(sys.argv[0])} "
            f"<absolute-folder-path> [output-json-path]"
        )
    folder = sys.argv[1]
    if not os.path.isabs(folder):
        sys.exit(f"Error: folder path must be absolute (got: {folder}).")
    if not os.path.isdir(folder):
        sys.exit(f"Error: not a directory: {folder}")

    output_path = sys.argv[2] if len(sys.argv) == 3 else os.path.join(folder, "results.json")

    files = list(iter_media_files(folder))

    print("Resemble AI -- Batch Deepfake Detection with Secure Upload\n")
    print("=" * 60)
    print(f"Folder:  {folder}")
    print(f"Output:  {output_path}")
    print(f"Files:   {len(files)}")
    print(f"Workers: {MAX_WORKERS}\n")

    if not files:
        print("No media files to process.")
        return

    succeeded: list[dict] = []
    failed: list[dict] = []

    with ThreadPoolExecutor(max_workers=MAX_WORKERS) as pool:
        future_to_path = {pool.submit(process_file, p): p for p in files}
        for future in as_completed(future_to_path):
            path = future_to_path[future]
            name = os.path.basename(path)
            try:
                record = future.result()
                succeeded.append(record)
                metrics = (record["detect"].get("metrics") or {})
                print(
                    f"  [OK]  {name}  "
                    f"label={metrics.get('label')}  score={metrics.get('aggregated_score')}"
                )
            except Exception as e:
                failed.append({"file": path, "error": str(e)})
                print(f"  [ERR] {name}  {e}")

    with open(output_path, "w") as f:
        json.dump(
            {"folder": folder, "succeeded": succeeded, "failed": failed},
            f,
            indent=2,
        )

    print()
    print("=" * 60)
    print(f"Done. {len(succeeded)} succeeded, {len(failed)} failed (of {len(files)}).")
    print(f"Results written to {output_path}")


if __name__ == "__main__":
    main()
